File: src/expressionparser/ExpressionParser.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package expressionparser;

import expressionparser.tokens.Value;
import expressionparser.tokenparsers.NumberTokenParser;
import expressionparser.tokenparsers.FromTokenableParser;
import java.util.ArrayList;

/**
 *
 * @author Lomztein
 */
public class ExpressionParser {

    private final ITokenParser[] parsers = new ITokenParser[]{
        new NumberTokenParser(),
        new FromTokenableParser(),};

    private final ArrayList<TokenData> tokens = new ArrayList<>();

    public double parse(String input) {

        input = removeWhitespace(input);

        parseTokens(input);

        while (tokens.size() > 1) {

            TokenData next = findNextToEvaluate(tokens);
            int nextIndex = tokens.indexOf(next);
            
            if (next == null)
                break;

            EvaluateResult result = ((IEvaluable) next.token).evaluate(tokens, nextIndex);

            int emptySpot = 0;
            for (int i = 0; i < result.spendTokenIndices.length; i++) {
                tokens.set(result.spendTokenIndices[i], null);
                emptySpot = result.spendTokenIndices[i];
            }
            
            tokens.set(emptySpot, new TokenData(new Value(result.result), next.depth));

            for (int i = 0; i < tokens.size(); i++) {
                if (tokens.get(i) == null) {
                    tokens.remove(i);
                    i--;
                }
            }

        }

        if (tokens.get(0).token instanceof IEvaluable) {
            return ((IEvaluable)tokens.get(0).token).evaluate(tokens, 0).result;
        }
        return ((Value) tokens.get(0).token).getValue();

    }

    private void parseTokens(String input) {

        System.out.println("Parsing: " + input);

        int balance = 0;
        for (int i = 0; i < input.length(); i++) {

            if (input.charAt(i) == '(') {
                balance++;
            }
            if (input.charAt(i) == ')') {
                balance--;
            }

            String substring = input.substring(i);
            for (ITokenParser parser : parsers) {
                ParseResult result = parser.parse(substring);
                if (result != null) {
                    tokens.add(new TokenData(result.token, balance));
                    i += result.getLength() - 1;
                }
            }

        }

    }

    private TokenData findNextToEvaluate(ArrayList<TokenData> tokens) {

        int highestPrecedence = Integer.MIN_VALUE;
        TokenData highestToken = null;

        for (int i = 0; i < tokens.size(); i++) {

            TokenData at = tokens.get(i);

            if (!(at.token instanceof IEvaluable)) {
                continue;
            }

            IEvaluable evaluable = (IEvaluable) at.token;

            int precedence = evaluable.getPrecedence() + (at.depth * 10);
            if (highestPrecedence < precedence) {
                highestToken = at;
                highestPrecedence = precedence;
            }

        }

        return highestToken;

    }

    private static String removeWhitespace(String input) {
        String newString = "";
        for (int i = 0; i < input.length(); i++) {

            char curChar = input.charAt(i);
            if (curChar != ' ') {
                newString += curChar;
            }

        }

        return newString;
    }

}

File: C#/MathematicalExpressionParser/ExpressionParser.cs
using Lomztein.MathematicalExpressionParser.Evaluation;
using Lomztein.MathematicalExpressionParser.Parsers;
using Lomztein.MathematicalExpressionParser.Token;
using System;
using System.Collections.Generic;

namespace Lomztein.MathematicalExpressionParser {

    public class ExpressionParser {

        public ITokenParser[] Parsers { get; set; } = new ITokenParser[]{
            new NumberTokenParser(),
            new FromTokenableParser(),
        };

        private List<Tuple<IToken, int>> Tokens { get; set; } = new List<Tuple<IToken, int>> ();

        public double Parse(String input) {

            input = RemoveWhitespace (input);
            ParseTokens (input);

            while (Tokens.Count > 1) {

                var next = FindNextToEvaluate ();
                int nextIndex = Tokens.IndexOf (next);

                if (next == null)
                    break;

                var result = ((IEvaluable)next.Item1).Evaluate (Tokens, nextIndex);

                int emptySpot = 0;
                for (int i = 0; i < result.spendIndicies.Length; i++) {
                    Tokens[result.spendIndicies[i]] = null;
                    emptySpot = result.spendIndicies[i];
                }

                Tokens[emptySpot] = new Tuple<IToken, int> (new Value (result.result), next.Item2);

                for (int i = 0; i < Tokens.Count; i++) {
                    if (Tokens[i] == null) {
                        Tokens.RemoveAt (i);
                        i--;
                    }
                }

            }

            if (Tokens[0].Item1 is IEvaluable evaluable) {
                return (evaluable.Evaluate (Tokens, 0)).result;
            }
            return ((Value)Tokens[0].Item1).Number;

        }

        private void ParseTokens(String input) {

            int balance = 0;
            for (int i = 0; i < input.Length; i++) {

                if (input[i] == '(') {
                    balance++;
                }
                if (input[i] == ')') {
                    balance--;
                }

                string substring = input.Substring (i);
                foreach (ITokenParser parser in Parsers) {

                    var result = parser.Parse (substring);

                    if (result.token != null) {
                        Tokens.Add (new Tuple<IToken, int> (result.token, balance));
                        i += result.result.Length - 1;
                    }

                }

            }

        }

        private Tuple<IToken, int> FindNextToEvaluate() {

            int highestPrecedence = int.MinValue;
            Tuple<IToken, int> highestToken = null;

            for (int i = 0; i < Tokens.Count; i++) {

                var at = Tokens[i];

                if (at.Item1 is IEvaluable evaluable) {

                    int precedence = evaluable.Precedence + (at.Item2 * 10);
                    if (highestPrecedence < precedence) {
                        highestToken = at;
                        highestPrecedence = precedence;
                    }
                }
            }

            return highestToken;

        }

    private static string RemoveWhitespace(String input) {
        string newString = "";
        for (int i = 0; i < input.Length; i++) {

            char curChar = input[i];
            if (curChar != ' ') {
                newString += curChar;
            }

        }

        return newString;
    }

}
}